src/libraries/System.Private.CoreLib/src/System/Reflection/ModifiedGenericType.cs

```diff
@@ -0,0 +1,47 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Reflection
+{
+    internal sealed partial class ModifiedGenericType : ModifiedType
+    {
+        private readonly ModifiedType[] _argumentTypes;
+
+        public ModifiedGenericType(
+            Type genericType,
+            object? signatureProvider,
+            int rootSignatureParameterIndex,
+            int nestedSignatureIndex,
+            int nestedSignatureParameterIndex,
+            bool isRoot)
+            : base(
+                  genericType,
+                  signatureProvider,
+                  rootSignatureParameterIndex,
+                  nestedSignatureIndex, // To support modifiers on generic types, pass nestedSignatureIndex+1
```

Comment:
I am concerned about testability -- this is the `List<const T>` case. I suppose I can use Emit in the test assembly, create a Type from it and go from there.

Reply:
You can add .IL helper test assembly. Libraries tests have number of those.

Reply:
It looks like open generic methods can have modifiers; I'll work on that.

```
.method public hidebysig instance void  MyMethodWithOpenGeneric<T>(!!T modopt([System.Runtime]System.Runtime.CompilerServices.IsConst) t) cil managed
```

Reply:
That also means that these modifiers can show up in base-classes (`Type.BaseType`), implemented interfaces (`Type.GetInterfaces`) and generic constraints (`Type.GetGenericParameterConstraints`). Do we need to override these methods on modified types?

For example, `class MyType<T> : Base<const int>, IFace<const uint> where T: IFace2<const byte>`.

Note that the `modopt` are used to encode `unmanaged` constraint in C#. For example, `class MyType<T> where T: unmanaged` uses modopt to encode `unmanaged`. I do not think there is a way to extract this information using reflection APIs today. This would make it possible.

If we do not want to do it now, we may want to change these methods to throw so that overriding them with the correct implementation later is not a breaking change.

The TypeDelegator makes these issues easy to miss. I am wondering whether it would be better to stop inheriting modified types from TypeDelegator and make all APIs that implicitly switch back to unmodified types throw instead.

Reply:
Clarification: I was able to add modifiers to generic parameters on an open generic method:
`Foo<T>(T modopt(IsConst) t)`
but I was not successful adding modifiers to the generic parameters on an open generic type.

Reply:
So unless I'm missing something on generic types + modifiers, this PR should be complete in this area.

Reply:
How about the `Type.BaseType` and other similar properties?

For example, if I write this in C#:
```
class T : List<delegate* unmanaged[Cdecl]<void>[]>
{
}
```
Do we want to make it possible to find out that the function pointer is Cdecl?

Reply:
> How about the Type.BaseType and other similar properties?

We support the original planned ones of course:
```cs
public virtual Type[] GetFunctionPointerCallingConventions();
public virtual Type[] GetFunctionPointerParameterTypes();
public virtual Type GetFunctionPointerReturnType();
public virtual Type[] GetOptionalCustomModifiers();
public virtual Type[] GetRequiredCustomModifiers();
```

Then I added support recently for:
```cs
public virtual Type[] GetGenericArguments();
```

Here's the rest; since they are not in the "signature" held by field\parameter\property these just fall through to `UnderlyingSystemType` meaning no modifiers:
```cs
public override Type? DeclaringType { get; }
public abstract Type? BaseType { get; }
public override Type? ReflectedType { get; }
public virtual Type GetEnumUnderlyingType();
public virtual Type[] GenericTypeArguments { get; }
public virtual Type[] GetGenericParameterConstraints();
public virtual Type GetGenericTypeDefinition();
public Type? GetInterface(string name);
public abstract Type[] GetInterfaces();
public Type? GetNestedType(string name);
public abstract Type? GetNestedType(string name, BindingFlags bindingAttr);
public Type[] GetNestedTypes();
public abstract Type[] GetNestedTypes(BindingFlags bindingAttr);
```

Reply:
For MetadataLoadContext, we could enable "modififed types" by default, which would allow modifiers to be obtained for the base type scenario, as well as everything else provided MetadataReader does the appropriate callback. I'll prototype this today.

Reply:
Based on offline discussion we should:
- Throw NSE on all members on the modified Type that may return, directly or indirectly, another Type. For example `BaseType` will throw.
- Keep MLC semantics in sync with the runtime. Throw NSE like the runtime and don't enable modified types automatically (doing that would cause ReferenceEquals() semantics to change; Equals() or == would have to be used instead).
- We have the option in the future to support `BaseType` and other properties as TBD and whether the modified Type tree can keep growing past Type declarations.